### PR TITLE
feat: add prettier as formatter

### PR DESCRIPTION
**Pull request by Galimede**
## What does this PR do?

* This PR introduces Prettier as a formatter for our codebase
* It also includes the update of Stylelint since Prettier will format the CSS and that Stylelint doesn't format anymore
* Note: ESLint was not updated here as the release was too recent and not every plugins are up to date yet
* Note: this PR also changed the Node version we use to the latest 

## How to review

* Check the commits
* npm install and use the new command : `npm run format` and see the changes by yourself locally
* Check that everything is working fine with the new Node version
* Don't hesitate to play with the Prettier config and especially with the `printWidth` rule
  * I've set it to 120 because we have pretty long lines and it was easier to read HTML in our components 
  * But feel free to play around with it while it's not fixed yet 

## TODO

* [x] Update files with the new order from Stylelint (can be done in a separate PR)
* [x] Apply the formatter to the files (can be done in a separate PR)
* [x] Add the `format:ci` to our GitHub action
* [x] Add the new npm scripts to the task doc

### CODE REVIEW

**Comment by florian-sanders-cc** [comment]
:wave: Hey there thanks for testing & setting up all of this @Galimede 

My only question is: could we open a new issue about sorting imports with prettier?
What do you think about [trivago/prettier-plugin-sort-imports](https://github.com/trivago/prettier-plugin-sort-imports)?
 
Reading through Hubert's feedback, I have one question:

> Who runs the format:ci?
> 
>     Should we add this in a GitHub action?
>     Maybe it should be in prepack?
>     Maybe in a precommit hook?

I'm not sure but I wonder if CI should format.
Formatting means files would be modified and these modifications would need to be commited no?
While it is possible to commit & push stuff from an action, I would rather use `prettier --check ...` to let the CI fail if formatting is missing so people can format and push formatting as part of their commits (not a commit dedicated to formatting I mean) before a PR is merged :thinking: 

TLDR; no `format:ci` & a `format:check-ci` instead?

File: docs/adr/adr-0026-introducing-prettier.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
+kind: '📌 Architecture Decision Records'
+---
+# ADR 0026: Introducing Prettier as a formatter
+
+🗓️ 2024-06-13 · ✍️ Mathieu Degand
+
+
+This ADR tries to explain why we decided to introduce Prettier while migrating our linting tools (ESLint/Stylelint) to their latest version.
```
Comment (pdesoyres-cc) [inline docs/adr/adr-0026-introducing-prettier.md:8]:
```suggestion
This ADR explains why we decided to introduce Prettier while migrating our linting tools (ESLint/Stylelint) to their latest version.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+## Context
+
+One year ago we made the decision to add Stylelint to our components library. You can find the how and why on the [ADR]().
```
Comment (pdesoyres-cc) [inline docs/adr/adr-0026-introducing-prettier.md:12]: question: is there a missing link here `[ADR]()` ?

File: docs/adr/adr-0026-introducing-prettier.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+- Biome is multipurpose, it has a formatter, a linter and more.
+
+#### Cons
+
+- It is opiniated, which means you can't really override things to match your need.
+- It only has some default agnostic formatting options.
+- While being multipurpose is a pro, it's also a con. If you want to use only one feature, you'll still need to get the whole package and deactivate the features you don't need.
+
+While we could have chosen to go for Biome as a formatter, and maybe go even further and choose it as a linter too, there's a few reasons while we decided not to go for this solution.
+- As a linter, there's not a lot of options and it can't check our CSS, so that's a clear no go.
```
Comment (pdesoyres-cc) [inline docs/adr/adr-0026-introducing-prettier.md:44]: nitpick can you add that it is because our CSS is inside JS files

File: docs/adr/adr-0026-introducing-prettier.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+[Prettier](https://prettier.io) is an opiniated code formatter.
+
+#### Pros
+
+- Can format both JS, CSS, template literals in one take.
+
+#### Cons
+
+- There's not a lot of options.
+  - It is justified by the `opiniated formatter` logic.
```
Comment (pdesoyres-cc) [inline docs/adr/adr-0026-introducing-prettier.md:75]: nitpick: Maybe one tick is enough.

```suggestion
- There's not a lot of options which is justified by the `opiniated formatter` logic.
```

**Review by pdesoyres-cc — APPROVED** [review]
I reviewed the ADR. Nice writing! I just have some nitpicks or questions.

File: docs/adr/adr-0026-introducing-prettier.md (hunk 2 above)
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:12]: **nitpick:** 
```suggestion
One year ago we made the decision to add Stylelint to our components library. You can find the how and why in the [ADR]().
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+One year ago we made the decision to add Stylelint to our components library. You can find the how and why on the [ADR]().
+At that time Stylelint was in version `14`ish and is now in version `16` as of this writing.
+During a meeting while discussing versions of the tools we use, we decided to take a look at the updates and the feasibility to update them to the latest version.
+And here came the problem: linters don't (want to) format anymore.
+
+That's the case for Stylelint, that decided to deprecate format rules in version `15`, while removing them all in the latest version (16).
+But that's also the case for ESLint that made the same decision.
+In fact most if not all of the linters tend to switch to not format the code now.
+
+Thus said we took the decision to handle the subject on our side and find a formatter before updating both tools. (ESLint and Stylelint)
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:21]: **nitpick:**
```suggestion
Thus said, we took the decision to handle the subject on our side and find a formatter before updating both ESLint and Stylelint.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+- Biome is multipurpose, it has a formatter, a linter and more.
+
+#### Cons
+
+- It is opiniated, which means you can't really override things to match your need.
+- It only has some default agnostic formatting options.
+- While being multipurpose is a pro, it's also a con. If you want to use only one feature, you'll still need to get the whole package and deactivate the features you don't need.
+
+While we could have chosen to go for Biome as a formatter, and maybe go even further and choose it as a linter too, there's a few reasons while we decided not to go for this solution.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:43]: **nitpick:**
```suggestion
While we could have chosen to go for Biome as a formatter, and maybe go even further and choose it as a linter too, there are a few reasons while we decided not to go for this solution.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+[Prettier](https://prettier.io) is an opiniated code formatter.
+
+#### Pros
+
+- Can format both JS, CSS, template literals in one take.
+
+#### Cons
+
+- There's not a lot of options.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:74]: **nitpick:**
```suggestion
- There are not a lot of options.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  - It is justified by the `opiniated formatter` logic.
+
+Prettier is the option we chose to use, you can find more information about the why on the dedicated section below.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:77]: **nitpick:**
```suggestion
Prettier is the option we chose to use, you can find more information about the why within the dedicated section below.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+- While we knew all these tools existed we wanted to know what Lit users were using
+  - Knowing that we asked the Lit discord's what they were using
+    - Someone told us that they were using the new `Stylistic` package for ESLint that keeps formatting rules
+    - They were also using a Prettier plugin for CSS
+    - While it seemed a good idea at first, you'll find why it wasn't adapted to our case below.
+
+## Investigation after Lit Discord's question (TODO: remove this)
+
+- I have asked the Lit Discord's to know if some users used specific tools to format their Lit components
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:100]: **nitpick:** I would stick with "We" since it's what we use everywhere and I don't think we ever go for "I" within ADRs but it's not very important so you decide of course :wink: 
```suggestion
- We have asked the Lit Discord to know if some users used specific tools to format their Lit components
```
or
```suggestion
- We have asked the Lit's Discord to know if some users used specific tools to format their Lit components
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  - It seems like they're using a `Stylistic` package that should maintains the deprecated rules
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:101]: **nitpick:**
```suggestion
  - It seems like they're using a `Stylistic` package that should maintain the deprecated rules
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+### Others
+
+- While we knew all these tools existed we wanted to know what Lit users were using
+  - Knowing that we asked the Lit discord's what they were using
+    - Someone told us that they were using the new `Stylistic` package for ESLint that keeps formatting rules
+    - They were also using a Prettier plugin for CSS
+    - While it seemed a good idea at first, you'll find why it wasn't adapted to our case below.
+
+## Investigation after Lit Discord's question (TODO: remove this)
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:98]: **nitpick:** You have a TODO here :eyes:

File: docs/adr/adr-0026-introducing-prettier.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+## Solution
+
+### Why did we choose Prettier?
+
+We decided to go for Prettier for various reasons:
+
+- It is widely used and maintained.
+- Has integration with IDEs.
+- It enforces code style. Beside our internal conventions we didn't have a proper formatter enforcing.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:121]: **nitpick:** 
```suggestion
- It enforces code style. Other than our internal conventions we didn't have a proper formatter enforcing.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+Note: As we didn't update ESLint yet, see details on the dedicated section below, we added the Prettier ESLint plugin so that ESLint formatting rules doesn't conflict with Prettier.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:123]: **nitpick:**
```suggestion
Note: As we haven't updated ESLint yet, see details on the dedicated section below, we've added the Prettier ESLint plugin so that ESLint formatting rules doesn't conflict with Prettier.
```

File: docs/adr/adr-0026-introducing-prettier.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+- Got rid of the `postcss-css-in-js` custom syntax as it's now deprecated.
+  - Decided to use the dedicated `postcss-lit` syntax instead.
+- Removed our idiomatic CSS properties order.
+  - We took this decision because the package was/will not be maintain(ed).
+  - We chose to use the `stylelint-order` package instead and use an alphabetical order.
+- Removed all the Stylelint formatting rules from our configuration file.
+
+### Why didn't we update ESLint (for now)?
+
+With the update of Stylelint and the introduction of Prettier, we also considered to update ESLint to version 9. However we decided after some tests to not do it yet.
```
Comment (florian-sanders-cc) [inline docs/adr/adr-0026-introducing-prettier.md:155]: **nitpick:**
```suggestion
With the update of Stylelint and the introduction of Prettier, we also considered to updating ESLint to version 9. However, after some testing, we have decided to hold it off for now.
```

**Review by florian-sanders-cc — CHANGES_REQUESTED** [review]
Gg @Galimede this is very clear and easy to understand :wink: 

I've left a few nitpicks but nothing really serious :+1:

File: docs/adr/adr-0026-introducing-prettier.md [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+At some point we considered going for the `stylistic` package for ESLint to keep our usage.
+While it would have been great, this wouldn't solve the lack of formatter for Stylelint.
+
+Considering this, we tried to use the `stylelint-prettier` plugin to format only our CSS.
+However, we couldn't do that because:
+`postcss-lit` uses babel and if you use that parser `stylelint-prettier` will just skip your file.
+So we tried to force Prettier to use the `css` parser instead for Stylelint but that resulted in some errors.
+
+Knowing that we couldn't use this plugin and thus not having a formatter for our Stylelint/CSS part we decided not to use Stylistic.
+Moreover, if we wanted to use both ESLint with Stylystic and prettier for the CSS it would have caused some conflicts.
```
Comment (roberttran-cc) [inline docs/adr/adr-0026-introducing-prettier.md:121]: **🔨 issue:**

```suggestion
Moreover, if we wanted to use both ESLint with Stylistic and prettier for the CSS it would have caused some conflicts.
```

**Review by roberttran-cc — APPROVED** [review]
Very nice (and awaited) job!

A minor typo but everything else is good to me!

**Review by florian-sanders-cc — APPROVED** [review]
LGTM, gg again for the ADR and all this work :clap:

**Review by HeleneAmouzou — APPROVED** [review]
I reviewed the ADR. It is very well explained, GG ! :clap:

File: src/components/cc-block-section/cc-block-section.js [hunk 16]
```diff
@@ -25,34 +24,34 @@ export class CcBlockSection extends LitElement {
...
+    => This can be done from the <cc-block> with CSS only ::slotted(cc-block-section).
+    REQUIREMENT #2: As the point is to add separation lines/borders, we don't want to style the first one.
+    => This cannot be done with ::slotted(cc-block-section:first-of-type).
+    SOLUTION: Add the margin directly in <cc-block-section> with :host(:not(:first-of-type))
+    WARNINGS #1: This breaks the "don't add margin on the :host,
+     do it on the parent" but we exlained why we cannot do it.
+    WARNINGS 21: This code will work even if the parent is not a <cc-block>.
+     We could do something to prevent this pb with connectedCallback
+     and a class or an attribute but it's a bit much just for this.
+ */
```
Comment (florian-sanders-cc) [inline src/components/cc-block-section/cc-block-section.js:50]: oups, not sure we can do anything about that but the indenting failed here :disappointed:
